storage/transfermanager/integration_test.go

```diff
@@ -17,160 +17,582 @@ package transfermanager
+			},
+			{
+				desc: "incorrect generation",
+				in: &DownloadObjectInput{
+					Bucket:     tb.bucket,
+					Object:     objectName,
+					Generation: &incorrectGen,
+				},
+				want: &DownloadOutput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Err:    storage.ErrObjectNotExist,
+				},
+			},
+			{
+				desc: "conditions: generationmatch",
+				in: &DownloadObjectInput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Conditions: &storage.Conditions{
+						GenerationMatch: r.Attrs.Generation,
+					},
+				},
+				want: &DownloadOutput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Attrs:  &r.Attrs,
+				},
+			},
+			{
+				desc: "conditions do not hold",
+				in: &DownloadObjectInput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Conditions: &storage.Conditions{
+						GenerationMatch: incorrectGen,
+					},
+				},
+				want: &DownloadOutput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Err:    errors.Join(&googleapi.Error{Code: 412}, status.Error(codes.FailedPrecondition, "")),
+				},
+			},
+			{
+				desc: "timeout",
+				in: &DownloadObjectInput{
+					Bucket: tb.bucket,
+					Object: objectName,
+				},
+				timeout: time.Nanosecond,
+				want: &DownloadOutput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Err:    context.DeadlineExceeded,
+				},
+			},
+			{
+				desc: "cancelled ctx",
+				in: &DownloadObjectInput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					ctx:    cancelledCtx,
+				},
+				timeout: time.Nanosecond,
+				want: &DownloadOutput{
+					Bucket: tb.bucket,
+					Object: objectName,
+					Err:    context.Canceled,
+				},
+			},
+		} {
+			t.Run(test.desc, func(t *testing.T) {
+				w := &testWriter{}
+
+				test.in.Destination = w
 
-		if got.Err != nil {
-			t.Errorf("result.Err: %v", got.Err)
-			continue
-		}
+				if test.in.ctx == nil {
+					test.in.ctx = ctx
+				}
 
-		if want, got := contentHashes[got.Object], writers[writerIdx].crc32c; got != want {
-			t.Fatalf("content crc32c does not match; got: %v, expected: %v", got, want)
-		}
+				got := test.in.downloadShard(c, test.timeout)
 
-		if got.Attrs.Size != objectSize {
-			t.Errorf("expected object size %d, got %d", objectSize, got.Attrs.Size)
+				if got.Bucket != test.want.Bucket || got.Object != test.want.Object {
+					t.Errorf("wanted bucket %q, object %q, got: %q, %q", test.want.Bucket, test.want.Object, got.Bucket, got.Object)
+				}
+
+				if diff := cmp.Diff(got.Attrs, test.want.Attrs); diff != "" {
+					t.Errorf("diff got(-) vs. want(+): %v", diff)
+				}
+
+				if !errorIs(got.Err, test.want.Err) {
+					t.Errorf("mismatching errors: got %v, want %v", got.Err, test.want.Err)
+				}
+			})
 		}
+	})
+}
+
+// errorIs is equivalent to errors.Is, except that it additionally will return
+// true if err and targetErr are googleapi.Errors with identical error codes,
+// or if both errors have the same gRPC status code.
+func errorIs(err error, targetErr error) bool {
```

Comment:
Clever idea, we should probably do something like this for other tests which force errors.